lib/core/util.c

```diff
@@ -58,58 +64,192 @@ static struct st_h2o_accept_data_t *create_accept_data(h2o_accept_ctx_t *ctx, h2
+    if (*conn == NULL) {
+        *conn = h2o_redis_create_connection(ctx->loop, sizeof(h2o_redis_conn_t));
+        (*conn)->on_connect = on_redis_connect;
+        (*conn)->on_close = on_redis_close;
+    }
+    return *conn;
+}
+
+#define BASE64_LENGTH(len) (((len) + 2) / 3 * 4 + 1)
+
+static h2o_iovec_t build_redis_key(h2o_iovec_t session_id, h2o_iovec_t prefix)
+{
+    h2o_iovec_t key;
+    key.base =  h2o_mem_alloc(prefix.len + BASE64_LENGTH(session_id.len));
+    if (prefix.len != 0) {
+        memcpy(key.base, prefix.base, prefix.len);
+    }
+    key.len = prefix.len;
+    key.len += h2o_base64_encode(key.base + key.len, session_id.base, session_id.len, 1);
+    return key;
+}
+
+static h2o_iovec_t build_redis_value(h2o_iovec_t session_data)
+{
+    h2o_iovec_t value;
+    value.base =  h2o_mem_alloc(BASE64_LENGTH(session_data.len));
+    value.len = h2o_base64_encode(value.base, session_data.base, session_data.len, 1);
+    return value;
+}
+
+#undef BASE64_LENGTH
+
+static void redis_resumption_on_get(redisReply *reply, void *_accept_data)
+{
+    struct st_h2o_accept_data_t *accept_data = _accept_data;
+    accept_data->redis_resumption_get_command = NULL;
+
+    h2o_iovec_t session_data;
+    if (reply != NULL && reply->type == REDIS_REPLY_STRING) {
+        session_data = h2o_decode_base64url(NULL, reply->str, reply->len);
+    } else {
+        session_data = h2o_iovec_init(NULL, 0);
+    }
+
+    h2o_socket_ssl_resume_server_handshake(accept_data->sock, session_data);
+
+    if (session_data.base != NULL)
+        free(session_data.base);
+}
+
+static void on_redis_resumption_get_failed(h2o_timeout_entry_t *timeout_entry)
+{
+    struct st_h2o_accept_data_t *accept_data = H2O_STRUCT_FROM_MEMBER(struct st_h2o_accept_data_t, resumption_failure_timeout, timeout_entry);
+    accept_data->redis_resumption_get_command = NULL;
+    h2o_socket_ssl_resume_server_handshake(accept_data->sock, h2o_iovec_init(NULL, 0));
+    h2o_timeout_unlink(timeout_entry);
+}
+
+static void redis_resumption_get(h2o_socket_t *sock, h2o_iovec_t session_id)
+{
+    struct st_h2o_accept_data_t *accept_data = sock->data;
+    h2o_redis_conn_t *conn = get_redis_connection(accept_data->ctx->ctx);
+
+    if (conn->state == H2O_REDIS_CONNECTION_STATE_CONNECTED) {
+        h2o_iovec_t key = build_redis_key(session_id, async_resumption_context.redis.prefix);
+        accept_data->redis_resumption_get_command = h2o_redis_command(conn, redis_resumption_on_get, accept_data, "GET %s", key.base);
+        free(key.base);
+    } else {
+        if (conn->state == H2O_REDIS_CONNECTION_STATE_CLOSED) {
+            // try to connect
+            h2o_redis_connect(conn, async_resumption_context.redis.host.base, async_resumption_context.redis.port);
+        }
+        // abort resumption
+        h2o_timeout_link(accept_data->ctx->ctx->loop, &accept_data->ctx->ctx->zero_timeout, &accept_data->resumption_failure_timeout);
```

Comment:
Would it be possible to reuse `st_accept_data_t::timeout` instead of introducing `resumption_failure_timeout`? I assume this is the only use-case for the variable.